### DISCUSSION TITLE
Using flux bootstrap, but preventing overwrite of an existing flux

### DISCUSSION THREAD

**Discussion by darrylweaver** [fluxcd/flux2#4235] [Proposals] [answered]
We are using flux bootstrap to install and configure an initial installation onto a brand new cluster.
We are using flux bootstrap to upgrade an already installed flux installation on an existing cluster.
This is great for those installations that are already managed via the Flux CLI command.

However, there are multiple implementations of Flux including Azure Flux AKS Add-on, Azure Arc Flux Add-on, the flux terraform provider to name just a few.

If flux was already installed and managed by a different source, such as via Azure Flux on AKS or with a cluster build using the flux terraform provider, then the Flux CLI should not take over management of the installation due to conflicts with the existing source.

It would therefore be beneficial for an option to the flux bootstrap command to safely detect an existing flux already installed and exit instead of upgrading the existing flux installation to handle this case.
The benefits to users would be to allow them to automate flux bootstrapping with confidence that any existing management of flux would not be disrupted, even accidentally.  

For example: `--no-overwrite` option would allow automation to be safe when attempting to bootstrap flux to a cluster only if it is not already installed without additional commands.

**Comment by stefanprodan** [comment, marked as answer]
When running `flux bootstrap` or `flux install` we should detect if Flux is already installed on the cluster with Helm (what the Azure Flux addon does). 

We can look for the `app.kubernetes.io/managed-by: Helm` label on the `gitrepositories.source.toolkit.fluxcd.io` CRD. If the CRD exists and has this label & value, we should prompt like so:

```
Flux <VERSION> has been installed on this cluster with Helm!
Are you sure you want to proceed with overriding the Helm installation? Y/N
```

The Flux version should be extracted from the `app.kubernetes.io/version` label value.

Both `flux bootstrap` and `flux install` commands should have a `--force` flag, when set, this will bypass the above prompt.

xref: https://github.com/fluxcd/flux2/issues/4342

---

To prevent people from upgrading by mistake a cluster with `flux install` when the cluster has been bootstrapped, we can look for the `kustomize.toolkit.fluxcd.io/name` and `kustomize.toolkit.fluxcd.io/namespace` labels on the `gitrepositories.source.toolkit.fluxcd.io` CRD. If the CRD exists and has these labels, we should return an error like so:

```
This cluster has been bootstrapped with Flux <VERSION>! Please use 'flux bootstrap' to upgrade.
```

The Flux version should be extracted from the `app.kubernetes.io/version` label value.

xref: https://github.com/fluxcd/flux2/issues/4341
